digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java

```diff
@@ -160,8 +160,13 @@ protected void runWithHeartbeat(TaskRequest request)
                         callback.retryTask(request, agentId, ex.getRetryInterval().get(), ex.getStateParams(cf).get(), ex.getError(cf));
                     }
                     else {
-                        logger.error("Task {} failed.\n{}", request.getTaskName(), formatExceptionMessage(ex));
-                        logger.debug("", ex);
+                        if (request.isCancelRequested()) {
+                            logger.warn("Task {} canceled.", request.getTaskName());
+                        }
+                        else {
+                            logger.error("Task {} failed.\n{}", request.getTaskName(), formatExceptionMessage(ex));
+                            logger.debug("", ex);
+                        }
                         // TODO use debug to log stacktrace here
                         callback.taskFailed(request, agentId, ex.getError(cf).get());  // TODO is error set?
```

Comment:
What if `runWithWorkspace` throws `TaskExecutionException` not from `operator.cleanup(mergedRequest)`? It looks the clean up won't be called.

Reply:
If `runWithWorkspace` throws `TaskExecutionException`, it depends on what type of `TaskExecutionException`, but it will retry the task if it is called by `TaskExecutionException#ofNextPolling` which is used by a polling-based task. And then `cleanup` be executed at that time if `mergedResest.isCancelRequested()` is true. 
On the other hand, if cleanup failed to perform, target components will remain, I added a note about that and throw `TaskExecutionException` after cleanup to stop the task. https://github.com/treasure-data/digdag/pull/1480/commits/afb261043d8778f5c77f8d05736e73578b6a47c8